File: PSI-2020-backend/PSI-2020-backend/Models/Database/SubjectCard.cs
using PSI_2020_backend.Models.Database.Bases;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Threading.Tasks;

namespace PSI_2020_backend.Models.Database
{
    public class SubjectCard : ModelBase
    {
        public int Number { get; set; }
        public bool IsDone { get; set; }
        public List<Program> Lectures { get; set; }
        public List<Program> Laboratories { get; set; }
        public List<Program> Projects { get; set; }
        public List<Program> Seminaries { get; set; }
        public List<Program> Exercises { get; set; }
        public string Prerequisites { get; set; }
        public string PrerequisitesEng { get; set; }
        public string Bibliography { get; set; }
        public string BibliographyEng { get; set; }
        public string Tools { get; set; }
        public string ToolsEng { get; set; }
        public string Aims { get; set; }
        public string AimsEng { get; set; }
        public List<string> Professors { get; set; }
        public List<LearningEffect> LearningEffects{ get; set; }
    }
}

File: frontend/src/app/models/subject-card.ts
import { LearningEffect } from "./learning-effect";
import { Program } from "./program";

export class SubjectCard {
    id: number;
    isDone: boolean = false;
    lectures: Program[] = [];
    laboratories: Program[] = [];
    projects: Program[] = [];
    seminaries: Program[] = [];
    exercises: Program[] = [];
    prerequisites: string;
    prerequisitesEng: string;
    bibliography: string;
    bibliographyEng: string;
    tools: string;
    toolsEng: string;
    aims: string;
    aimsEng: string;
    professors: string[] = [];
    learningEffects: LearningEffect[];

    constructor(init?: Partial<SubjectCard>) {
        Object.assign(this, init);
    }

    static create()
    {
        return new SubjectCard({
            isDone: false,
            lectures: [],
            laboratories: [],
            projects: [],
            seminaries: [],
            exercises: [],
            professors: [],
            learningEffects: []
        });
    }

    static hasUniqueTitleError(card: SubjectCard)
    {
        return (!card.isDone && !(
            new Set(card.laboratories.map(p => p.subject)).size == card.laboratories.length &&
            new Set(card.lectures.map(p => p.subject)).size == card.lectures.length &&
            new Set(card.exercises.map(p => p.subject)).size == card.exercises.length &&
            new Set(card.projects.map(p => p.subject)).size == card.projects.length &&
            new Set(card.seminaries.map(p => p.subject)).size == card.seminaries.length));
    }
}
